src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -202,21 +216,116 @@ public boolean isOpen()
       public void close() throws IOException
       {
         open = false;
+        useDelegateSmooshWriter = false;
         internalFiles.put(name, new Metadata(currOut.getFileNum(), startOffset, currOut.getCurrOffset()));
 
         if (bytesWritten != currOut.getCurrOffset() - startOffset) {
-          throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
-        }
-        if (bytesWritten != size) {
-          throw new IOException(
-              String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+            throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
+          }
+          if (bytesWritten != size) {
+            throw new IOException(
+                String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
+            );
+          }        
+        if(!mergeOnProgress && !useDelegateSmooshWriter)
+        {
+        	mergeWithSmoosher();
         }
+
       }
     };
   }
+  
+  public OutputStream makeOutputStream(String filename) throws IOException
+  {
+    File retFile = createdFiles.get(filename);
+    if (retFile == null) {
+      retFile = File.createTempFile("smoosher", filename);
+      retFile.deleteOnExit();
+      createdFiles.put(filename, retFile);
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } else  {
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } 
+  }
+  
+  public InputStream makeInputStream(String filename) throws IOException
+  {
+    final File retFile = createdFiles.get(filename);
+    return retFile == null ? null : new FileInputStream(retFile);
+  }
+  
+  private synchronized void mergeWithSmoosher() throws IOException
```

Comment:
You made this synchronized in order to try to make this object thread-safe and protect from concurrent modifications.  But you also added state to the FileSmoosher that is being modified but not protected.  Either this object is thread-safe and you protect everything or it's not thread-safe and you expect the user to protect against concurrent interactions.